### PR TITLE
[WIP] Add Atlas - Retrieval Augmented Language Model

### PR DESCRIPTION
**Pull request by ae99**
# What does this PR do?

Implements Atlas: Few-shot Learning with Retrieval Augmented Language Model as mentioned here
 https://github.com/huggingface/transformers/issues/20503

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten, @lhoestq, @patil-suraj
cc  @patrick-s-h-lewis and @gizacard

### CODE REVIEW

**Comment by ae99** [comment]
This branch is very much a WIP currently, but for anyone interested here is roughly how I plan to structure things, aiming to roughly mesh the shape of the original implementation with Transformer's existing patterns. For the most part, I hope to make its usage as similar as possible to `T5ForConditionalGeneration`.

This is all new to me, so any feedback would be super helpful!

```python
class AtlasConfig():
    pass

class AtlasTrainer(Trainer):
    pass

class AtlasPreTrainedModel(PreTrainedModel):
    pass

class AtlasModel(AtlasPreTrainedModel):
    def __init__(self, queryPassageEncoder, reader, retriever):
        self.queryPassageEncoder = queryPassageEncoder # UntiedDualEncoder
        self.reader = reader # FiD
        self.retriever = retriever # HFIndexBase

class FiD(T5ForConditionalGeneration):
    def __init__(self):
        self.encoder = FiDStack()
        self.decoder = FiDStack()

class FiDStack(T5Stack):
    pass

class UntiedDualEncoder(torch.nn.Module):
    def __init__(self, query_contriever, passage_contriever):
        self.query_contriever = query_contriever
        self.passage_contriever = passage_contriever

class Contriever(BertModel):
    pass

class HFIndexBase():
    pass

class AtlasRetriever:
    def __init__(self, index):
        self.index = index # HFIndexBase
```

---

The existing RAG implementation makes its sub-models easily swappable, however, the inputs and outputs expected by 'reader' model (the name given to the T5 encoder/decoder in the original implementation) here are non-standard due to the fusion-in-decoder mechanism, so I don't plan to make these models as easily swappable as I think that would complicate things unnecessarily.

As I'm not doing this, it seems it may be best practice to copy implementation (w/ "Copied from" comments) of models like the BertModel and T5ForConditionalGeneration rather than import - if that's the case I'll switch these across once the PR's almost ready.

---

There is some complexity here in how we make the model trainable E2E within Huggingface's patterns, which I haven't yet looked into deeply. I wonder whether a `class AtlasTrainer(Trainer)` would make sense, which can implement the various continuous re-indexing strategies described in the original paper.

**Comment by sgugger** [comment]
Yes, please do use the approach of copying model code and adding `# Copied from` comments as it's more inline with the general approach in the library (RAG being an exception :-) )

**Comment by sgugger** [comment]
cc @ArthurZucker

**Comment by patrickvonplaten** [comment]
@ArthurZucker @ae99 let me know if you need help with anything - think this is a super cool addition!

**Comment by ae99** [comment]
> @ArthurZucker @ae99 let me know if you need help with anything - think this is a super cool addition!

Hey @patrickvonplaten and @ArthurZucker! I think the general structure of this model is mostly in place. I'd love to get an early review on the PR from you just to check if things are looking ok and confirm the major things are roughly fitting patterns correctly.

I have a few temporary notebooks `save_pretrained.ipynb`, `test_retriever.ipynb` and `test_model.ipynb` in place of actual tests at the moment if you would like to get a sense of usage. Like RAG I have a dedicated retriever, but I've cut this down to mostly be a small wrapper around a dataset+index for now. Documentation and tests haven't been touched at all yet, and everything is very WIP still!

**Comment by akashe** [comment]
Hi @ae99, I would also like to contribute. Let me know if there is something I can help you with.

**Comment by patrickvonplaten** [comment]
@ArthurZucker could you maybe take a look here? :-) Let me know if you need some help

**Comment by patrickvonplaten** [comment]
@akashe feel free to give this PR a review as well if you'd like to help a bit :-)

**Comment by ArthurZucker** [comment]
Will review now 😉

File: docs/source/en/model_doc/atlas.mdx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+The abstract from the paper is the following:
+
+*<INSERT PAPER ABSTRACT HERE>*
+
+Tips:
+
+<INSERT TIPS ABOUT MODEL HERE>
+
+This model was contributed by [INSERT YOUR HF USERNAME HERE](https://huggingface.co/<INSERT YOUR HF USERNAME HERE>).
+The original code can be found [here](<INSERT LINK TO GITHUB REPO HERE>).
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/atlas.mdx:29]: Same here. Just flagging as a TODO

File: docs/source/en/model_doc/atlas.mdx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+-->
+
+# Atlas
+
+## Overview
+
+The Atlas model was proposed in [<INSERT PAPER NAME HERE>](<INSERT PAPER LINK HERE>) by <INSERT AUTHORS HERE>.
+<INSERT SHORT SUMMARY HERE>
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/atlas.mdx:18]: This needs to be filled. Also make sure you run `make repo-consistency` as it seems that the model was not added to all of our `README.mdx` though this should be automatic 😉

File: forward_pass_mocking.ipynb [hunk 3]
```diff
@@ -0,0 +1,1446 @@
+{
```
Comment (ArthurZucker) [inline forward_pass_mocking.ipynb:1]: We don't add notebooks especially not here! Is there a particular reason for this one?

File: save_pretrained.ipynb [hunk 4]
```diff
@@ -0,0 +1,370 @@
+{
```
Comment (ArthurZucker) [inline save_pretrained.ipynb:1]: Same here

File: src/transformers/__init__.py [hunk 5]
```diff
@@ -1879,6 +1880,9 @@
     _import_structure["models.rag"].extend(
         ["RagModel", "RagPreTrainedModel", "RagSequenceForGeneration", "RagTokenForGeneration"]
     )
+    _import_structure["models.atlas"].extend(
+        ["AtlasModel", "AtlasPreTrainedModel", "AtlasSequenceForGeneration", "AtlasTokenForGeneration"]
```
Comment (ArthurZucker) [inline src/transformers/__init__.py:1884]: These model names are a bit strange. Might be inspired from rag but we dropped these kind of notation. Also from reading the paper it seems that you need `AtlasForConditionalGeneration` (as it follows the T5 formatting of prompts) ans `AtalsForMaskedLM` for training/fine tuning the model. This comment is valid for the rest of the `__init__` files and of course the modelling

File: src/transformers/models/atlas/__init__.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+try:
+    if not is_torch_available():
+        raise OptionalDependencyNotAvailable()
+except OptionalDependencyNotAvailable:
+    pass
+else:
+    _import_structure["modeling_atlas"] = [
+        "AtlasModel",
+        "AtlasPreTrainedModel",
+    ]
```
Comment (ArthurZucker) [inline src/transformers/models/atlas/__init__.py:38]: The init is missing the other models that you define in the `modeling` file.

File: src/transformers/models/atlas/configuration_atlas.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+            decoder_start_token_id=decoder_start_token_id,
+            forced_eos_token_id=forced_eos_token_id,
+            is_encoder_decoder=is_encoder_decoder,
+            prefix=prefix,
+            vocab_size=vocab_size,
+            **kwargs,
+        )
+        assert (
+            "retriever" in kwargs and "generator" in kwargs
+        ), "Config has to be initialized with retriever and generator config"
```
Comment (ArthurZucker) [inline src/transformers/models/atlas/configuration_atlas.py:120]: No assert, let's raise an appropriate error

File: src/transformers/models/atlas/configuration_atlas.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        retriever_config = kwargs.pop("retriever")
+        retriever_model_type = retriever_config.pop("model_type")
+        decoder_config = kwargs.pop("generator")
+        decoder_model_type = decoder_config.pop("model_type")
+
+        from ..auto.configuration_auto import AutoConfig
```
Comment (ArthurZucker) [inline src/transformers/models/atlas/configuration_atlas.py:126]: No import in the middle of the function

File: src/transformers/models/atlas/configuration_atlas.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+        self.retriever = AutoConfig.for_model(retriever_model_type, **retriever_config)
+        self.generator = AutoConfig.for_model(decoder_model_type, **decoder_config)
```
Comment (ArthurZucker) [inline src/transformers/models/atlas/configuration_atlas.py:129]: These are configs so probably `retriever_config` and `generator_config`

File: src/transformers/models/atlas/retriever.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (ArthurZucker) [inline src/transformers/models/atlas/retriever.py:1]: We have a 1 model 1 file policy so this should be part of the `modeling_file`.  There should also be no dependencies with other model, but rather a use of the `# Copied from` statements. See [here](https://huggingface.co/blog/transformers-design-philosophy)

File: src/transformers/models/atlas/tokenization_atlas.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+from ...tokenization_utils_base import BatchEncoding
+from ...utils import logging
+from .configuration_atlas import AtlasConfig
+
+
+logger = logging.get_logger(__name__)
+
+
+class AtlasTokenizer:
```
Comment (ArthurZucker) [inline src/transformers/models/atlas/tokenization_atlas.py:28]: We have to use the super class `PreTrainedTokenizer` to fit in the library. The function that you have to re-define are included.

File: tests/models/atlas/test_modeling_atlas.py [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+        AutoConfig,
+        AutoModel,
+        AutoModelForSeq2SeqLM,
+        DPRContextEncoder,
+        AtlasConfig,
+        AtlasModel,
+        AtlasRetriever,
+        AtlasSequenceForGeneration,
+        AtlasTokenForGeneration,
+        AtlasTokenizer,
```
Comment (ArthurZucker) [inline tests/models/atlas/test_modeling_atlas.py:55]: Not all of these models require torch

File: tests/models/atlas/test_modeling_atlas.py [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+from transformers.utils import cached_property, is_datasets_available, is_faiss_available, is_torch_available
+
+TOLERANCE = 1e-3
+
+T5_SAMPLE_VOCAB = get_tests_dir("fixtures/test_sentencepiece.model")
+if is_torch_available() and is_datasets_available() and is_faiss_available():
+    import torch
+    from datasets import Dataset
+
+    import faiss
```
Comment (ArthurZucker) [inline tests/models/atlas/test_modeling_atlas.py:44]: Let's remove this dependency and test against raw values in the test file

File: tests/models/atlas/test_tokenization_atlas.py [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+
+
+if is_torch_available() and is_datasets_available() and is_faiss_available():
+    from transformers.models.atlas.configuration_atlas import AtlasConfig
+    from transformers.models.atlas.tokenization_atlas import AtlasTokenizer
+
+
+@require_faiss
+@require_torch
+class AtlasTokenizerTest(TestCase):
```
Comment (ArthurZucker) [inline tests/models/atlas/test_tokenization_atlas.py:37]: Same for every test, we have to use `TokenizerTesterMixin, unittest.TestCase`  (`from ...test_tokenization_common import TokenizerTesterMixin
`

**Review by ArthurZucker — COMMENTED** [review]
Hey, thanks for the effort to start this huge work! There are a lot of things to update w.r.t the inspiration. I would maybe advise to try using  `transformers-cli add new model like` . Tell me if you need more help in designing ( I know composed models can be pretty challenging, don't hesitate)

File: src/transformers/models/atlas/modeling_atlas.py [hunk 15]
```diff
@@ -0,0 +1,380 @@
...
+
+            query_emb = self.retriever(input_ids=query_input_ids, attention_mask=query_attention_mask, is_passages=False)
+            retriever_tokens = {k: v.reshape(-1, v.size(-1)) for k, v in retriever_tokens.items()}
+
+            passage_emb = self.retriever(**retriever_tokens, is_passages=True).to(query_emb)
+            passage_emb = passage_emb.view(bsz, -1, passage_emb.size(-1))
+            retriever_score = torch.einsum("id, ijd->ij", [query_emb, passage_emb])
+
+            if decoder_input_ids is None:
+                decoder_input_ids = self.generator._shift_right(labels)
```
Comment (akashe) [inline src/transformers/models/atlas/modeling_atlas.py:280]: Labels are usually the final answers expected from the model. In the [Atlas implementation](https://github.com/facebookresearch/atlas/blob/4b6df19687210e4c8e9df2795664195041f9e45c/src/atlas.py#L229) the decoder_input_ids inlcude the query also.

File: src/transformers/models/atlas/modeling_atlas.py [hunk 16]
```diff
@@ -0,0 +1,380 @@
...
+        if retriever_index is None:
+            raise ValueError("retriever_index is None, from_pretrained not implemented yet.")
+
+        self.retriever_index = retriever_index
+        self.retriever = retriever
+        self.generator = generator
+
+    # todo
+    # - make it possible to separate the retrieval and generation steps
+    def forward(
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/atlas/modeling_atlas.py:213]: If we can improve on RAG, it's  also super welcome 😉

**Comment by akashe** [comment]
Hey @ae99 , are you still working on the integration? If not then let me know, I would be happy to continue from where you left.

**Comment by ae99** [comment]
> Hey @ae99 , are you still working on the integration? If not then let me know, I would be happy to continue from where you left.

Hey @akashe, that'd be perfect.

**Comment by ArthurBook** [comment]
Hi!
This one would be really relevant for something that we are working on in my org.
What's the status on it? We may be able to chip in.

**Comment by ArthurZucker** [comment]
Hey! We have not really picked it up, if the community needs it we can probably come back to it, but I would advise to just put the model on the hub following this [tutorial](https://huggingface.co/docs/transformers/custom_models)! 🤗
